### PR TITLE
chore(aws_s3 sink): refactor the S3 sink

### PR DESCRIPTION
**Pull request by tobz**
In the same vein as #8825, this PR is a refactor of the S3 sink.

Instead of the traditional sink model where all pertinent behavior -- batching, encoding, etc -- is stacked up and nested together _outside_ of the sink itself, we move it inside the sink and switch from a push to a pull model.

This work is part of an ongoing investigation into Vector stalling when certain components are used together, due to what is likely faulty hand-rolled state machines in various components added to Vector over the years.  This is all about simplicity, and reducing the surface area where the implementation details of traits can have a fundamental impedance mismatch.

Feature checklist:
- [x] batching
- [x] partitioning
- [x] encoding
- [x] compression
- [x] auth
- [x] healthcheck

Signed-off-by: Toby Lawrence <toby@nuclearfurnace.com>

### CODE REVIEW

**Comment by jszwedko** [comment]
Kicked off integration tests here: https://github.com/timberio/vector/actions/runs/1201366426

**Comment by blt** [comment]
Alright, after re-kicking the integration tests the following fail:

```
    sinks::aws_s3::tests::integration_tests::acknowledges_failures
    sinks::aws_s3::tests::integration_tests::s3_gzip
    sinks::aws_s3::tests::integration_tests::s3_healthchecks_invalid_bucket
    sinks::aws_s3::tests::integration_tests::s3_insert_message_into
    sinks::aws_s3::tests::integration_tests::s3_insert_message_into_object_lock
```

File: Cargo.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ split-debuginfo = "unpacked" # Faster debug builds on macOS
 # This results in roughly a 5% reduction in performance when compiling locally vs when
 # compiled via the CI pipeline.
 [profile.release]
-debug = false     # Do not include debug symbols in the executable.
+debug = true    # Do not include debug symbols in the executable.
```
Comment (bruceg) [inline Cargo.toml:26]: Was this supposed to go into this PR? If so, the comment needs to be fixed.

Comment (blt) [reply Cargo.toml:26]: This does seem like a mistake to me.

Comment (blt) [reply Cargo.toml]: Resolved in d9c2a09f5

**Comment by blt** [comment]
New integration suite: https://github.com/timberio/vector/actions/runs/1210899260

File: src/sources/socket/mod.rs [hunk 2]
```diff
@@ -320,6 +320,7 @@ mod test {
     }
 
     #[tokio::test]
+    #[ignore] // see issue #9046
```
Comment (jszwedko) [inline src/sources/socket/mod.rs:323]: Could we drop this ignore now that it is fixed on master?

File: Cargo.toml [hunk 3]
```diff
@@ -128,6 +128,7 @@ tracing-tower = { git = "https://github.com/tokio-rs/tracing", default-features
 tracing-opentelemetry = { version = "0.15.0", default-features = false }
 opentelemetry = { version = "0.16.0", default-features = false, features = ["trace", "rt-tokio"] }
 opentelemetry-datadog = { version = "0.4.0", default-features = false, features = ["reqwest-client"] }
+console-subscriber = { git = "https://github.com/tokio-rs/console" }
```
Comment (jszwedko) [inline Cargo.toml:131]: Nit: should we do the same `default-features = false` here?

Also, do we still want this? I see the code that was using it is currently commented out.

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 4]
```diff
@@ -11,6 +11,7 @@ use leveldb::database::{
 };
 use std::collections::VecDeque;
 use std::fmt::Display;
+use std::future::Future;
```
Comment (jszwedko) [inline lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:14]: Could / should these changes be separated out into another PR?

I see there is a bunch of stale commented code as well.

File: src/sinks/aws_cloudwatch_logs/mod.rs [hunk 5]
```diff
@@ -74,7 +74,7 @@ pub struct CloudwatchLogsSinkConfig {
     #[serde(default)]
     pub batch: BatchConfig,
     #[serde(default)]
-    pub request: TowerRequestConfig<Option<usize>>,
+    pub request: TowerRequestConfig,
```
Comment (jszwedko) [inline src/sinks/aws_cloudwatch_logs/mod.rs:77]: This is related to #9007 I think.

File: src/trace.rs [hunk 6]
```diff
@@ -58,7 +58,11 @@ pub fn init(color: bool, json: bool, levels: &str, enable_datadog_tracing: bool)
     // This is a hidden and undocumented functionality.
     let metrics_layer_enabled = metrics_layer_enabled();
 
+    //let (tasks_layer, tasks_server) = console_subscriber::TasksLayer::new();
+    //tokio::spawn(tasks_server.serve());
+
     let subscriber = tracing_subscriber::registry::Registry::default()
+        //        .with(tasks_layer)
```
Comment (jszwedko) [inline src/trace.rs:65]: Do we want to keep the console subscriber? Or drop this commented code?

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 7]
```diff
@@ -108,14 +114,51 @@ where
...
+            loop {
+                match this.pending_read.take() {
+                    None => {
+                        // We have no pending read in-flight, so queue one up.
+                        let db = Arc::clone(&this.db);
+                        let read_offset = this.read_offset;
+                        let handle = tokio::task::spawn_blocking(move || {
+                            db.iter(ReadOptions::new())
+                                .from(&Key(read_offset))
+                                .take(1000)
```
Comment (jszwedko) [inline lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:126]: Not a big deal, but just noticing that this changed from 100 to 1000. Is that intentional?

File: src/sinks/aws_s3/service.rs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        };
+
+        Box::pin(async move {
+            let result = client.put_object(request).in_current_span().await;
+
+            // TODO: This is fine for testing, but we should have a better
+            // pattern for this.
+            emit!(S3EventsSent {
+                byte_size: body_len,
+            });
```
Comment (jszwedko) [inline src/sinks/aws_s3/service.rs:111]: Is this TODO still relevant?

File: src/sinks/aws_s3/sink.rs [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+                    // We could push this down to the I/O task if we wanted to.
+                    let request = build_request(key, batch, &self.options);
+                    if io_tx.send(request).await.is_err() {
+                        // TODO: change this to "error! + return Err" after
+                        // initial testing/debugging
+                        trace!(
+                            "Sink I/O channel should not be closed before sink itself is closed."
+                        );
+                        panic!("boom 1")
+                    }
```
Comment (jszwedko) [inline src/sinks/aws_s3/sink.rs:137]: Still relevant?

File: src/sinks/aws_s3/sink.rs [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+                }
+                // Partitioning failed for one or more events.
+                //
+                // TODO: this might be where we would insert something like the
+                // proposed error handling/dead letter queue stuff; events that
+                // _we_ can't handle, but some other component may be able to
+                // salvage
```
Comment (jszwedko) [inline src/sinks/aws_s3/sink.rs:144]: When would this fail in practice? Currently we drop any events that cannot be templated.

File: src/sinks/aws_s3/sink.rs [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+        .as_ref()
+        .cloned()
+        .unwrap_or_else(|| options.compression.extension().into());
+    let key = format!("{}/{}.{}", key, filename, extension);
+
+    // Process our events. This does all of the necessary encoding rule
+    // application, as well as encoding and compressing the events.  We're
+    // handed back a tidy `Bytes` instance we can send directly to S3.
+    //
+    // TODO: we need to do something with these
```
Comment (jszwedko) [inline src/sinks/aws_s3/sink.rs:295]: 👀

**Review by jszwedko — COMMENTED** [review]
Nice! I was generally able to follow the code more easily than the old implementation. I'll let @lukesteensen and @bruceg leave more informed comments though. There does seem to be a fair bit of TODOs left in here that we should probably resolve or drop.

File: src/sources/socket/mod.rs (hunk 2 above)
Comment (blt) [reply src/sources/socket/mod.rs:323]: Ah yeah, thank you.

Comment (blt) [reply src/sources/socket/mod.rs]: Resolved in 8552aab07

File: src/sinks/aws_s3/sink.rs [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+                    .map(move |result| {
+                        logic.update_finalizers(result, finalizers);
+
+                        // If the rx end is dropped we still completed
+                        // the request so this is a weird case that we can
+                        // ignore for now.
+                        let _ = tx.send((seqno, batch_size));
+                    })
+                    .instrument(info_span!("request", request_id = %seqno));
+                tokio::spawn(fut);
```
Comment (lukesteensen) [inline src/sinks/aws_s3/sink.rs:211]: It's a little confusing that this is the "IO task", but it seems like we're spawning all of the actual IO off into independent tasks. I could be missing something, but it makes me wonder if it's worthwhile to have this extra task split out with the accompanying channel hop.

File: src/sinks/aws_s3/sink.rs [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+                        _ => None,
+                    }),
+                // TODO: We don't ship metrics to the S3, but if we did, would this be right? or is
+                // there an actual field we should be checking similar to above?
+                Event::Metric(_) => Some(Utc::now()),
+            })
+            .unwrap_or_else(|| Utc::now());
+        let formatted_ts = last_event_ts.format(&time_format);
+        */
+        let formatted_ts = Utc::now().format(options.filename_time_format.as_str());
```
Comment (lukesteensen) [inline src/sinks/aws_s3/sink.rs:274]: This seems very reasonable and likely preferable to the old commented-out logic above. I'd say we document the change and delete the old version above.

File: src/sinks/aws_s3/sink.rs [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+            }
+        }
+    }
+
+    // Build our compressor first, so that we can encode directly into it.
+    let mut writer = {
+        // This is a best guess, because encoding could add a good chunk of
+        // overhead to the raw, in-memory representation of an event, but if
+        // we're compressing, then we should end up net below the capacity.
+        let buffer = Vec::new();
```
Comment (lukesteensen) [inline src/sinks/aws_s3/sink.rs:348]: The comment makes it sound like we should be using `Vec::with_capacity` here.

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs (hunk 7 above)
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:126]: Sans comment I couldn't tell you. I would assume this is either an artifact of testing but whether it's a tag-along or an intentional change I don't know.

File: src/trace.rs (hunk 6 above)
Comment (blt) [reply src/trace.rs:65]: I'll remove it. I do think it ended up being handy, but I'd like to see it properly introduced in a separate PR.

Comment (blt) [reply src/trace.rs]: Resolved in 0f4aac626

File: Cargo.toml (hunk 3 above)
Comment (blt) [reply Cargo.toml]: Resolved in 0f4aac626

**Review by lukesteensen — COMMENTED** [review]
Overall this makes a lot of sense! Good to see us moving away from the old `futures::Sink` stack towards something more manageable.

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs (hunk 4 above)
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:14]: Not having written this myself I'm concerned about pulling this work out into a separate PR. I know both the s3 changes and the disk buffer changes were validated together and if we were to separate it all out this needed change would get pushed back another day probably.

I can easily clear out the stale code though.

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 15]
```diff
@@ -108,14 +114,51 @@ where
...
+                                .collect::<Vec<_>>()
+                        });
+
+                        //trace!("queued up blocking read at offset={} (in-flight={}, total-reads={})",
+                        //    this.read_offset, this.in_flight_reads, this.total_reads);
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:131]: I think anything here that's commented out I will just drop. Seems like an artifact of the research Toby was doing.

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 16]
```diff
@@ -135,6 +178,7 @@ where
             // There are no writers left
             Poll::Ready(None)
         } else {
+            //trace!("going to sleep");
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs]: Resolved in 0aa71d395

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 17]
```diff
@@ -108,14 +114,51 @@ where
...
+                    Some(mut handle) => {
+                        match Pin::new(&mut handle).poll(cx) {
+                            Poll::Ready(r) => {
+                                match r {
+                                    Ok(items) => this.buffer.extend(items),
+                                    Err(e) => error!("error during read: {}", e),
+                                }
+
+                                this.pending_read = None;
+                                //this.in_flight_reads -= 1;
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs]: Resolved in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 18]
```diff
@@ -108,14 +114,51 @@ where
...
+                                .collect::<Vec<_>>()
+                        });
+
+                        //trace!("queued up blocking read at offset={} (in-flight={}, total-reads={})",
+                        //    this.read_offset, this.in_flight_reads, this.total_reads);
+
+                        // Store the handle, and let the loop come back around.
+                        this.pending_read = Some(handle);
+                        //this.in_flight_reads += 1;
+                        //this.total_reads += 1;
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs]: Resolved in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs]: Resolved in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs [hunk 19]
```diff
@@ -76,6 +78,10 @@ where
     pub(crate) max_uncompacted_size: usize,
     /// Last time that compaction was triggered.
     pub(crate) last_compaction: Instant,
+    // Pending read from the LevelDB datasbase
+    pub(crate) pending_read: Option<JoinHandle<Vec<(Key, Vec<u8>)>>>,
+    //pub(crate) in_flight_reads: usize,
+    //pub(crate) total_reads: usize,
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs]: Resolved in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs (hunk 4 above)
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs:14]: Dead code removed in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/mod.rs [hunk 20]
```diff
@@ -135,6 +135,9 @@ where
             unacked_sizes: VecDeque::new(),
             buffer: VecDeque::new(),
             last_compaction: Instant::now(),
+            pending_read: None,
+            //in_flight_reads: 0,
+            //total_reads: 0,
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/mod.rs]: Resolved in 0aa71d3

File: src/sinks/aws_s3/sink.rs [hunk 21]
```diff
@@ -0,0 +1,521 @@
...
+                Some(key) => {
+                    // We could push this down to the I/O task if we wanted to.
+                    let request = build_request(key, batch, &self.options);
+                    if io_tx.send(request).await.is_err() {
+                        // TODO: change this to "error! + return Err" after
+                        // initial testing/debugging
+                        trace!(
+                            "Sink I/O channel should not be closed before sink itself is closed."
+                        );
+                        panic!("boom 1")
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/sink.rs:136]: Lol

File: lib/vector-core/buffers/src/disk/leveldb_buffer/writer.rs [hunk 22]
```diff
@@ -122,6 +122,7 @@ where
                     // stall.
                     //
                     // To avoid this we notify the reader to notify this writer.
+                    //trace!("waking up reader (caught up case)");
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/writer.rs]: Resolved in 0aa71d3

File: lib/vector-core/buffers/src/disk/leveldb_buffer/writer.rs [hunk 23]
```diff
@@ -190,6 +191,7 @@ where
         self.writebatch = Writebatch::new();
         self.batch_size = 0;
         self.write_notifier.wake();
+        //trace!("waking up reader");
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/vector-core/buffers/src/disk/leveldb_buffer/writer.rs]: Resolved in 0aa71d3

File: src/sinks/aws_cloudwatch_logs/mod.rs (hunk 5 above)
Comment (blt) [reply src/sinks/aws_cloudwatch_logs/mod.rs:77]: Ah yeah, seems likely.

File: src/sinks/aws_s3/service.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/service.rs:111]: Yeah, considering it doesn't properly know how the request went it'll for sure double-emit on retries. I'm inclined to leave this as-is, comment included, and compare experiences with this PR and #8825 when Toby comes back from vacation. Then we can start the process of trying to normalize the two implementations to one another, spin off tickets for the TODOs etc.

File: src/sinks/aws_s3/partitioner.rs [hunk 24]
```diff
@@ -0,0 +1,33 @@
+use crate::{internal_events::TemplateRenderingFailed, template::Template};
+use vector_core::{event::Event, partition::Partitioner};
+
+/// Partitions items based on the generated S3 object key for the given event.
+///
+/// TODO: Realistically, this could be a generic "template partitioner", since
+/// I'm guessing other sinks might want to partition based on a key generated
+/// from event data.
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/partitioner.rs:8]: I do think this could be pulled up but I'm inclined to not open an issue yet, per my reasoning [here](https://github.com/timberio/vector/pull/8884/files#r703858426).

File: src/sinks/aws_s3/service.rs (hunk 8 above)
Comment (blt) [reply src/sinks/aws_s3/service.rs:111]: Please see https://github.com/timberio/vector/pull/8884/files#r703858426

File: src/sinks/aws_s3/sink.rs [hunk 25]
```diff
@@ -0,0 +1,521 @@
...
+            .expect("same sink should not be run twice");
+        let partitioner = self
+            .partitioner
+            .take()
+            .expect("same sink should not be run twice");
+
+        let io = run_io(io_rx, service, acker).in_current_span();
+        let _ = tokio::spawn(async move {
+            io.await;
+        });
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/sink.rs:114]: Oh yeah, I think it could be.

[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/sink.rs]: Resolved in 07b0b336b

File: src/sinks/aws_s3/sink.rs (hunk 13 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:272]: Agreed. Where do we have the current behavior documented?

File: src/sinks/aws_s3/sink.rs (hunk 11 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:293]: I'm honestly not sure what this is referring to. Feels vestigial.

Comment (blt) [reply src/sinks/aws_s3/sink.rs]: Resolved in de3c2340b

File: src/sinks/aws_s3/sink.rs (hunk 14 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:344]: Indeed. Right now it's a best-guess of 0.

Comment (blt) [reply src/sinks/aws_s3/sink.rs]: Bumped in 69a60138f

File: src/sinks/aws_s3/sink.rs (hunk 10 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:142]: It will fail if the template references a field that doesn't exist in the log, seems to me. The partitioner currently kicks out a `TemplateRenderingFailed` so we don't drop this silently (even if the message ends up being non-local to this code).

File: src/sinks/aws_s3/sink.rs (hunk 9 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs]: Resolved in 5b90ea0bd

File: src/sinks/aws_s3/sink.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/sink.rs]: Resolved in 5b90ea0bd

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (blt) [reply Cargo.toml]: I don't believe it was. Now removed.

File: src/sinks/aws_s3/sink.rs (hunk 13 above)
Comment (lukesteensen) [reply src/sinks/aws_s3/sink.rs:270]: It looks like we're not very specific currently. Source is [here](https://github.com/timberio/vector/blob/f1132d045b4ed0bcedc8983bc775af98bcba3b6d/website/cue/reference/components/sinks/aws_s3.cue#L323-L356). Likely worth specifying that it's the time the request is made and maybe dropping a note in the changelog. The only case I can think of where this would be a meaningful change is when working with historical data or similar.

File: src/sinks/aws_s3/sink.rs (hunk 12 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:207]: That's a good question. Absent any specific measure of this code I can't say, but I'll leave this sub-conversation unresolved so it's obvious for later discussion. 

/cc @tobz

File: src/sinks/aws_s3/sink.rs [hunk 26]
```diff
@@ -0,0 +1,521 @@
...
+                }
+                // Partitioning failed for one or more events.
+                //
+                // TODO: this might be where we would insert something like the
+                // proposed error handling/dead letter queue stuff; events that
+                // _we_ can't handle, but some other component may be able to
+                // salvage
+                None => {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/sink.rs:143]: This is covered by the partitioner, I believe.

File: src/sinks/aws_s3/sink.rs (hunk 13 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:270]: I think, if your amenable to the notion, I'd like to make a follow-up issue for this. We should have a test in place that enforces documented behavior, which we don't right now.

File: src/sinks/aws_s3/partitioner.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/partitioner.rs:8]: https://github.com/timberio/vector/issues/9078

File: src/sinks/aws_s3/service.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/service.rs:111]: I'll leave this sub-conversation unresolved so it's obvious for later discussion.

**Review by jszwedko — APPROVED** [review]
🚀

**Review by blt — APPROVED** [review]
I like this. In #8825 I've tried to push everything into stream processing which ends up being awkward. I think you sidestep that nicely here by spawning tasks and communicating over queues. I would still like to come up with a way of allowing folks to write a sink in a stream-y way without needing to spawn up their own tasks and what not but still saturate all the cores they have but that's something for later. 

I also appreciate that you re-use the existing service building stack here, which I think I'll crib for #8825. Having a default stack that users pop a `Service` into seems ideal from a maintenance point of view.

File: src/sinks/aws_s3/sink.rs (hunk 10 above)
Comment (tobz) [reply src/sinks/aws_s3/sink.rs:140]: Exactly.

File: src/sinks/aws_s3/service.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/sinks/aws_s3/service.rs:111]: Yeah, my inclination here is that this is probably a good use case for a piece of Tower middleware that can utilize the same machinery we use to drive retries vis a vis detecting if a response is good or not, and then emitting the appropriate metrics.  This doesn't actually need to be an S3-specific metric as is, since we only know it is by the `component_type` tag anyways.

[the remark this answers is not in the document]
Comment (blt) [reply src/sinks/aws_s3/service.rs:111]: 100%

File: src/sinks/aws_s3/sink.rs (hunk 12 above)
Comment (tobz) [reply src/sinks/aws_s3/sink.rs:207]: So we spawn the HTTP call, yes, but the I/O task is the one responsible for _driving_ the I/O.  It's not I/O in the literaly sense of sending and receiving the bytes over the wire, but tracking the request future and dealing with acknowledging the messages (in order) when the responses are resolved.

I modeled it this way as the start of trying to introduce more concurrency into the process.  The plan in my head was that it also probably makes sense to do encoding/compression in the I/O task.

Comment (blt) [reply src/sinks/aws_s3/sink.rs:207]: FWIW I have the Datadog logs work staged out so that encoding/compression takes place in a separate part of the pipeline from IO.

File: src/sinks/aws_s3/sink.rs (hunk 14 above)
Comment (tobz) [reply src/sinks/aws_s3/sink.rs]: Agreed.  The code was `with_capacity` when I wrote it, because I had access to the raw size, in bytes, of the events in the batch.  Our more generic code for `Batcher` doesn't currently expose that, so I had to change to the line as it stands, but left the comments with the plan of going back and augmenting the batch type to expose the total bytes.

File: src/sinks/aws_s3/sink.rs (hunk 12 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs:207]: Which is to say, I think it's a good idea. I'll crib ideas from this PR for that work so we'll get to see how it works out in practice in a different sink this week.

File: src/sinks/aws_s3/sink.rs (hunk 14 above)
Comment (blt) [reply src/sinks/aws_s3/sink.rs]: That's an excellent idea. Do you have an issue open for it already?
